### PR TITLE
Add support for Chinese (Traditional)

### PR DESCRIPTION
**Pull request by pan93412**
This PR distinguishes Simplified Chinese and Traditional Chinese[^1], and adds the translation, quoting, and partial support of numbering.

1. I renamed the constant `CHINESE` to `CHINESE_SIMPLIFIED` to distinguish them.
2. The language code of `TRAD_CHINESE` is named `zht` – the only 3-char code in `Lang`.
3. The numbering feature can't be implemented as we can't distinguish translations of the 1st number in `zht` and `zhs`. It is better to add a “lang” flag to indicate which numbering kind should be used. I now mark it as `#[allow(unused)]` and add a `TODO` on it.
4. For the translation of both new language code, I propose translating `CHINESE_SIMPLIFIED` as `中文（简体）`; `CHINESE_TRADITIONAL` as `中文（繁體）`.
 
[^1]: Traditional Chinese is used mainly by people in Taiwan, Hong Kong, etc.

### CODE REVIEW

File: library/src/layout/table.rs [hunk 1]
```diff
@@ -288,7 +288,8 @@ impl LocalName for TableElem {
         match lang {
             Lang::ARABIC => "جدول",
             Lang::BOKMÅL => "Tabell",
-            Lang::CHINESE => "表",
+            Lang::SIMP_CHINESE => "表",
+            Lang::TRAD_CHINESE => "表",
```
Comment (reknih) [inline library/src/layout/table.rs:292]: simp has a few questionable implications in English.
```suggestion
            Lang::CHINESE_SIMPLIFIED => "表",
            Lang::CHINESE_TRADITIONAL => "表",
```

File: library/src/meta/numbering.rs [hunk 2]
```diff
@@ -260,6 +260,13 @@ enum NumberingKind {
     Symbol,
     Hebrew,
     Chinese,
+    // TODO: Pick the numbering pattern based on languages choice.
+    // As the `1st` numbering character of Chinese (Simplifed) and
+    // Chinese (Traditional) is same, we are unable to determine
+    // if the context is Simplified or Traditional by only this
+    // character.
+    #[allow(unused)]
+    TradChinese,
```
Comment (reknih) [inline library/src/meta/numbering.rs:269]: This should be `ChineseSimplified` and `ChineseTraditional` as well, to make it consistent with the other enum.

File: library/src/meta/numbering.rs [hunk 3]
```diff
@@ -437,14 +445,18 @@ impl NumberingKind {
                 }
                 fmt
             }
-            Self::Chinese => {
+            l @ (Self::Chinese | Self::TradChinese) => {
                 let chinesecase = match case {
                     Case::Lower => ChineseCase::Lower,
                     Case::Upper => ChineseCase::Upper,
                 };
```
Comment (reknih) [inline library/src/meta/numbering.rs:452]:
```suggestion
                let chinese_case = match case {
                    Case::Lower => ChineseCase::Lower,
                    Case::Upper => ChineseCase::Upper,
                };
```

File: library/src/text/quotes.rs [hunk 4]
```diff
@@ -146,6 +146,8 @@ impl<'s> Quotes<'s> {
             "bs" | "fi" | "sv" => ("’", "’", "”", "”"),
             "hu" | "pl" | "ro" => ("’", "’", "„", "”"),
             "ru" | "no" | "nb" | "nn" | "ua" => ("’", "’", "«", "»"),
+            // we prefer “「『』」” classically
```
Comment (reknih) [inline library/src/text/quotes.rs:149]: Thanks for the comment, but your code already states that :)

File: src/doc.rs [hunk 5]
```diff
@@ -524,7 +524,17 @@ pub struct Lang([u8; 3], u8);
 impl Lang {
     pub const ARABIC: Self = Self(*b"ar ", 2);
     pub const BOKMÅL: Self = Self(*b"nb ", 2);
-    pub const CHINESE: Self = Self(*b"zh ", 2);
+
+    #[deprecated = "use Lang::SIMP_CHINESE instead"]
+    pub const CHINESE: Self = Self::SIMP_CHINESE;
```
Comment (reknih) [inline src/doc.rs:529]: We don't care about compiler breakage -- feel free to remove this constant

File: src/doc.rs [hunk 6]
```diff
@@ -524,7 +524,17 @@ pub struct Lang([u8; 3], u8);
...
+    /// Simplified Chinese.
+    // TODO: change to `zhs` for next breaking version
+    //       to distinguish from another Chinese:
+    //       `zht` (traditional Chinese)
```
Comment (reknih) [inline src/doc.rs:533]: Same here, feel free to apply the change

File: tests/typ/layout/quote-cjk.typ [hunk 7]
```diff
@@ -0,0 +1,6 @@
+
+---
+#text(lang: "zh", region: "CN")["你好, '世界'."]
+
+---
+#text(lang: "zh", region: "TW")["你好, '世界'."]
```
Comment (reknih) [inline tests/typ/layout/quote-cjk.typ:6]: Nice!

**Review by reknih — CHANGES_REQUESTED** [review]
Thank you for your translation. I added a few comments on enum and constant names and our current handling of breaking changes.

I also want to keep `zh` in addition to the explicit `zhs` and `zht` and make it pick the correct language variant for `zh` based on which region the user picked. Do you think that could work?

File: src/doc.rs (hunk 6 above)
Comment (pan93412) [reply src/doc.rs:533]: I'm not sure if it should be applied boldly – it may break documents using `#text(lang: 'zh')` :(

**Comment by pan93412** [comment]
> Thank you for your translation. I added a few comments on enum and constant names and our current handling of breaking changes.
> 
> I also want to keep `zh` in addition to the explicit `zhs` and `zht` and make it pick the correct language variant for `zh` based on which region the user picked. Do you think that could work?

I prefer to leverage our already-have `region` property, so we won't touch our `Lang` enumeration. How do you think? 

We may need to refactor `LocalName` trait to allow passing `region` into this method.

**Comment by peng1999** [comment]
The documentation says `lang` is a ISO 639-1/2/3 language code. It seems that `zhs` is not a valid code.

**Comment by szdytom** [comment]
> > I prefer to leverage our already-have region property
> 
> I'm second to this. The documentation says `lang` is a ISO 639-1/2/3 language code. It seems that `zhs` is not a valid code.

tthe standard code will be `zh-Hans` and `zh-Hant`. But that is going to a complex problem, actually there a at least 4 mainly used variants:

- `zh-Hans-CN` (or `zh-CN` for short): Simplified Chinese(Mandarin Chinese). mainly used in China mainland.
- `zh-Hant-TW` (or `zh-TW` for short): Traditional Chinese(Taiwanese), used in Taiwan province.
- `zh-Hant-HK` (or `zh-HK` for short): Traditional Chinese(Cantonese), used in Hongkong district.
- `zh-Hans-SG` (or `zh-SG` for short): mainly Simplified Chinese, but partly mixed Traditional Chinese, which are subject to the official regulations of Singapore(Singaporean), used in Singapore

**Comment by peng1999** [comment]
> tthe standard code will be `zh-Hans` and `zh-Hant`. But that is going to a complex problem, actually there a at least 4 mainly used variants:

It is not that complicated. We can already specify `lang-region` pair in current typst, and let's leverage it to support zh-cn and zh-tw first. More can be added later.

**Comment by pan93412** [comment]
I'm currently working on the LocalName refactor. PR of this will be landed soon.

File: tests/typ/layout/quote-cjk.typ [hunk 8]
```diff
@@ -0,0 +1,6 @@
+
+---
+#text(lang: "zh")["你好, '世界'."]
```
[the remark this answers is not in the document]
Comment (pan93412) [reply tests/typ/layout/quote-cjk.typ:3]: Sure, I'll remove the related changes!

**Pull request closed without merging**